pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+            .checked_add(1)
+            .ok_or(ArithmeticError::Overflow)?;
+        RegisteredUsers::<T>::insert(user_account, user_payment_streams_count);
+
+        // Store the new payment stream in the PaymentStreams mapping
+        // We initiate the last_valid_proof and last_charged_proof with the current block number to be able to keep track of the
+        // time passed since the payment stream was originally created
+        PaymentStreams::<T>::insert(
+            bsp_id,
+            user_account,
+            PaymentStream {
+                rate,
+                last_valid_proof: frame_system::Pallet::<T>::block_number(),
+                last_charged_proof: frame_system::Pallet::<T>::block_number(),
+            },
+        );
+
+        Ok(())
+    }
+
+    /// This function holds the logic that checks if a payment stream can be updated and, if so, updates the payment stream in the PaymentStreams mapping.
+    pub fn do_update_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+        new_rate: BalanceOf<T>,
+    ) -> DispatchResult {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Ensure that the new rate is not 0 (should use remove_payment_stream instead)
+        ensure!(new_rate != Zero::zero(), Error::<T>::UpdateRateToZero);
+
+        // Check that a payment stream between that BSP and user exists
+        ensure!(
+            PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamNotFound
+        );
+
+        // Get the information of the payment stream
+        let payment_stream = PaymentStreams::<T>::get(bsp_id, user_account)
+            .ok_or(Error::<T>::PaymentStreamNotFound)?;
+
+        // Verify that the new rate is different from the current one
+        ensure!(
+            payment_stream.rate != new_rate,
+            Error::<T>::UpdateRateToSameRate
+        );
+
+        // Charge the payment stream with the old rate before updating it to prevent abuse
+        let amount_charged = Self::do_charge_payment_stream(&bsp_account, user_account)?;
+        if amount_charged > Zero::zero() {
+            // We emit a payment charged event only if the user had to pay for its storage before the payment stream was updated
+            Self::deposit_event(Event::<T>::PaymentStreamCharged {
+                user_account: user_account.clone(),
+                backup_storage_provider_id: bsp_id,
+                amount: amount_charged,
+            });
+        }
+
+        // Check that the user is not flagged as without funds (after charging, to make sure it makes sense to update the payment stream)
+        ensure!(
+            !UsersWithoutFunds::<T>::contains_key(user_account),
```

Comment:
If the user is flagged as a consequence of the charging you did before, this is going to revert that. So in practice you wouldn't charge the user, nor flagged him, because the whole transaction would revert.

Reply:
Absolutely, don't know how I missed the revertion. Ty